File: DFS递归/leetcode_1022_sumRootToLeaf.cpp
#include <iostream>
#include <vector>
#include <functional>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    int dfs(TreeNode* node, int cur){
        if ((!node->left) && (!node->right)){
            return cur;
        }
        int res = 0;
        if (node->left){
            res += dfs(node->left, cur<<1 | node->left->val);
        }
        if (node->right){
            res += dfs(node->right, cur<<1 | node->right->val);
        }
        return res;
    }
public:
    int sumRootToLeaf(TreeNode* root) {
        return dfs(root, root->val);
    }
};

int main(){
    vector<int> tot = {1,0,1,0,1,0,1};
    function<TreeNode*(int)> dfs = [&](int indx){
        TreeNode* cur = new TreeNode(tot[indx]);
        if (indx*2+1 < tot.size() && tot[indx*2+1]>=0){
            cur->left = dfs(indx*2+1);
        }
        if (indx*2+2 < tot.size() && tot[indx*2+2]>=0){
            cur->right = dfs(indx*2+2);
        }
        return cur;
    };
    TreeNode* root = dfs(0);
    Solution s;
    cout << s.sumRootToLeaf(root) << endl;
}

File: 01.DFS递归/leetcode_1022_sumRootToLeaf.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        def dfs(node, cur):
            if not node.left and not node.right:
                return cur
            res = 0
            if node.left:
                res += dfs(node.left, cur<<1 | node.left.val)
            if node.right:
                res += dfs(node.right, cur<<1 | node.right.val)
            return res
        return dfs(root, root.val)

if __name__ == "__main__":
    s = Solution()
    tot = [1,0,1,0,1,0,1]
    def build(indx):
        cur = TreeNode(tot[indx])
        if indx*2+1 < len(tot) and tot[indx*2+1] != None:
            cur.left = build(indx*2+1)
        if indx*2+2 < len(tot) and tot[indx*2+2] != None:
            cur.right = build(indx*2+2)
        return cur
    root = build(0)
    print(s.sumRootToLeaf(root))
